In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
june = pd.read_pickle('../data/june.pkl')
june.head()

pubdatetime  latitude  longitude  \
0 2019-06-01 00:00:12   36.1202   -86.7534   
1 2019-06-01 00:00:12   36.1630   -86.7765   
2 2019-06-01 00:00:12   36.1202   -86.7533   
3 2019-06-01 00:00:12   36.1201   -86.7530   
4 2019-06-01 00:00:12   36.1622   -86.7806   

                                        sumdid  chargelevel  companyname  
0  Powered93627c35-0f62-5b81-a78d-75a4a92ecf47         90.0            5  
1  Powered17715097-e8a0-5494-a5ab-9b625796607d         63.0            5  
2  Powerede5cb95ae-b091-5a93-86fa-ededd946d0d7         77.0            5  
3  Powered71fa5e4f-1e17-54c4-936d-330df38cc2fa          0.0            5  
4  Poweredfa549dd6-40bb-5757-ac87-2c2528f2bc68          2.0            5

In [4]:
june.tail()

pubdatetime   latitude  longitude                sumdid  \
28046090 2019-06-30 23:59:55.273  36.125567 -86.776982  PoweredZYNSVOII2QSLD   
28046091 2019-06-30 23:59:55.273  36.121717 -86.775207  PoweredXOPAIWGL7SF62   
28046092 2019-06-30 23:59:55.273  36.127046 -86.777262  PoweredBYKFOWELGPKH6   
28046093 2019-06-30 23:59:55.273  36.123775 -86.790115  PoweredIAAQ7WODYXON7   
28046094 2019-06-30 23:59:55.273  36.121735 -86.775148  Powered3HJPPUNVYGL45   

          chargelevel  companyname  
28046090        100.0            3  
28046091         93.0            3  
28046092         61.0            3  
28046093         35.0            3  
28046094         96.0            3

In [5]:
june.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28037408 entries, 0 to 28046094
Data columns (total 6 columns):
 #   Column       Dtype         
---  ------       -----         
 0   pubdatetime  datetime64[ns]
 1   latitude     float64       
 2   longitude    float64       
 3   sumdid       object        
 4   chargelevel  float64       
 5   companyname  int64         
dtypes: datetime64[ns](1), float64(3), int64(1), object(1)
memory usage: 1.5+ GB


#note: there are 8 million more rows of data in June than there are in May.  The month is actually shorter than may, so it's not that ... Maybe more logging per scooter?  maybe more scooters in total?

In [6]:
june['sumdid'].value_counts() 

PoweredUNORTWSTRM4AM                           8631
PoweredBV42LPYJGVBL4                           8630
Powered9d204daf-7897-5dd5-8273-c2454b226d97    8628
Powered0286418d-3aff-5080-b4ee-a1ac8be225b8    8628
Poweredec02467e-2602-5e91-a74d-33f6f53a0c07    8628
                                               ... 
Powered2WZRP                                      1
Powered255293                                     1
PoweredUKDFR5O5HBMBO                              1
Powered911DI                                      1
PoweredXCPDB                                      1
Name: sumdid, Length: 7822, dtype: int64

nope - fewer scooters.  there were 8303 in may, there's 7822 in june.
I still see the top number of entries is about the same, and the bottom is still 1

In [7]:
june['companyname'].value_counts()

5    8482893
3    6152504
0    4215302
1    3084229
4    2507991
2    1830453
6    1764036
Name: companyname, dtype: int64

interesting - in may, company 3 had the most entries.  in june, it's company 5.  they stepped up their data collection.  do they have more scooters too?  Even though the overall number is smaller?

In [9]:
june_company_group = june.groupby("companyname")

In [10]:
june_scooter_company = june_company_group.agg({"sumdid": "nunique"})
print(june_scooter_company)

             sumdid
companyname        
0              2910
1              1152
2               223
3              1455
4               713
5              1036
6               333


company 0 dropped 154 scooters;
company 1 dropped 224 scooters;
company 2 same;
company 3 dropped 176 scooters;
company 4 gained 26 scooters;
company 5 gained 60 scooters;
company 6 dropped 13 scooters

In [12]:
june.isnull().sum()

pubdatetime      0
latitude         0
longitude        0
sumdid           0
chargelevel    285
companyname      0
dtype: int64

In [13]:
june_nullcharge_company= june.chargelevel.isnull().groupby(june.companyname).sum()
print(june_nullcharge_company)

companyname
0      0
1      0
2      0
3      0
4      0
5      0
6    285
Name: chargelevel, dtype: int64


i was careful to make sure that the company name/numbers were the same across all 3 months.  May all nulls were in company 4.   june, all nulls are in company 6.  what changed?

just for kicks, let's see how many single value entries there are in june ... 

In [14]:
june.groupby("sumdid").filter(lambda x: len(x) == 1)

pubdatetime   latitude  longitude                sumdid  \
3045     2019-06-01 00:03:27.257  36.157671 -86.773311  PoweredNIBNX6HK6AP5L   
639001   2019-06-01 16:40:58.923  36.082328 -86.725584          Powered911DI   
4372676  2019-06-05 13:02:07.010  36.147920 -86.740430        Powered8112481   
11366205 2019-06-12 20:58:57.477  36.137113 -86.800817  PoweredUKDFR5O5HBMBO   
16974731 2019-06-18 22:16:16.873  36.213138 -86.735270          Powered2WZRP   
18752982 2019-06-20 17:12:50.910  36.213265 -86.735226          PoweredXCPDB   
23512054 2019-06-26 03:50:10.600  36.149604 -86.797419        Powered6668448   
28043947 2019-06-30 23:56:15.457  36.177585 -86.750157         Powered255293   

          chargelevel  companyname  
3045             31.0            3  
639001           78.0            0  
4372676           0.0            4  
11366205         20.0            3  
16974731         62.0            0  
18752982         48.0            0  
23512054         92.0            4  
28043947         91.0            1

just 8 this time.  off hand, looks like different scooters  similar distribution across companies, dates, locations ... 